### ISSUE TITLE
Performance enhancemement opportunity in CPDF

### ISSUE THREAD

**Issue by GoogleCodeExporter** [jkoo7yeha/dompdf#291]
```
What would you like dompdf to do:
improve performance

Do you have an example:
I did some analysis using xdebug and wincachegrind on Vista/wamp (test 
environment - its faster ).
I am generating a relatively simple table with 6 columns and 20 rows.
It takes 1500ms to run on test system.

Using wincachegrind I can see that calls to 'file_exists' take up 319ms of the 
1501ms totaltime or ~21% of execution time.

I tracked the issue to '/lib/class.pdf.php'; class CPDF; function openFont; 
lines 2019, 2023, 2036 & 2044 primarily.
e.g.if ($this->isUnicode && !file_exists("$dir/$metrics_name")) { $metrics_name 
= $name . '.afm'; }
                             ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ 

In my test 'file_exists' was called 6,271 times, presumably just to verify that 
the same font/file exists!

PHP manual says that the results of 'file_exists' caches the output so should 
be insignificant in normal usage.

I would suggest that calling this function 6,271 times for such a small table 
is an extremely inefficient design, from a performance perspective.

I am not in a position to propose the solution as I am not up to speed with 
CPDF internals. However, perhaps checking that all font files exist - just 
once-  before processing would remove the need to check it every time. 
Alternatively, creating a new function which cached the results in a global 
array might also help. I supect reducing the need to call the function so many 
times will be the best solution.

Hope this helps :)
```

Original issue reported on code.google.com by `lukecjco...@gmail.com` on 26 May 2011 at 7:43

**Comment by GoogleCodeExporter** [comment]
```
Hi
I have investigated further and found that problem was caused by a 
configuration issue on my windows Vista dompdf install.

Once the configuration was fixed the performance issue went away. However, it 
did point out a possible problem, as follows: In class CPDF, function openFont 
the first piece of code uses the linux directory '/' separator to test the font 
file string. In my misconfigued case the string contained the windows '\' 
separator and this code does not cater for this situation.

I also noticed that there was possibly a typo in this function: there are 2 
calls to file_exists in this function with similar parameters one of which 
should be incorrect
     first one  -> !file_exists("$dir/$metrics_name")
                                     ^
     second one ->  file_exists($dir . $metrics_name)
It seems to me that the '/' is required or not, but should be the same in both 
cases?


So if this situation occurs legitamtely the test will fail and introduce an 
excessive performance overhead.

In conclusion, I guess this may be a non event in a properly configured system 
so  Issue 291 may be ignored/closed
```

Original comment by `lukecjco...@gmail.com` on 26 May 2011 at 10:22

**Comment by GoogleCodeExporter** [comment]
```
This is clearly something we should investigate, file access is something we 
should do less, even with the "stat" cache.
```

Original comment by `fabien.menager` on 27 May 2011 at 8:40
- Changed state: **Accepted**

**Comment by GoogleCodeExporter** [comment]
```
This issue was closed by revision r399.
```

Original comment by `fabien.menager` on 30 May 2011 at 8:56
- Changed state: **Fixed**

**Comment by GoogleCodeExporter** [comment]
Original comment by `eclecticgeek` on 30 May 2013 at 5:15
- Added labels: **Restrict-AddIssueComment-Commit**
